### ISSUE TITLE
[Q] Logging model weights in every epoch

### ISSUE THREAD

**Issue by najwalb** [wandb/wandb#6278]
I use artifacts to log model weights. Some of my runs crush (e.g. because they run out of time in SLURM), and I still want to log the last model weights as artifacts. I don't want to have an artifact version per epoch though, as this seems wasteful and some of my runs are really large (>1000 epochs). Is there anyway I can overwrite the model weights after every epoch, and guarantee they get logged to wandb even if my training loop does not finish?

Here are a few things I tried:

- Modifying the artifact directly:
```
# error: cannot modify a logged artifact
with wand.init() as run:
    for e in range(epochs):
        train(model)
        torch.save(model.state_dict(), 'model_latest.pt') 
        artifact = wandb.Artifact('model_latest', type='model')
        artifact.add_file(filename, name=filename)
        run.log_artifact(artifact, aliases=['latest-model'])
        if e>0:
             artifact.remove_file(filename, name=filename)
             artifact.add_file(filename, name=filename)
             run.log_artifact(artifact, aliases=['latest-model'])   
```
- Using upsert:
```
# error: upsert requires distributed runs
with wand.init() as run:
    for e in range(epochs):
        train(model)
        torch.save(model.state_dict(), 'model_latest.pt') 
        artifact = wandb.Artifact('model_latest', type='model')
        artifact.add_file(filename, name=filename)
        run.upsert_artifact(artifact, aliases=['latest-model'])
```
- Using the API to [change the artifact version from a previous version](https://docs.wandb.ai/guides/artifacts/create-a-new-artifact-version#create-a-new-artifact-version-from-an-existing-version):
```
# error: artifact does not have attribute 'new_draft()'
with wand.init() as run:
    for e in range(epochs):
        train(model)
        torch.save(model.state_dict(), 'model_latest.pt') 
        artifact = wandb.Artifact('model_latest', type='model')
        artifact.add_file(filename, name=filename)
        run.upsert_artifact(artifact, aliases=['latest-model'])
        if e>0:
            client = wandb.Api(project='myproject')
            saved_artifact = client.artifact("my_artifact:latest")
            draft_artifact = saved_artifact.new_draft()
            draft_artifact.remove("modified_file.txt")
            draft_artifact.add_file("modified_file.txt")
```

**Comment by ArtsiomWB** [comment]
Hi @najwalb, thank you for writing in! 

To double check: you are interested in being able to override your artifact every single epoch you are logging, instead of creating a new artifact version every single artifact.log?

Alongside, with that, you would like to upload your last artifact to wandb even after your run crashes?

**Comment by najwalb** [comment]
Hi @ArtsiomWB 

Yes, this is exactly it. The last point is optional as long as I am able to overwrite the artifacts logged in epochs.

**Comment by umakrishnaswamy** [comment]
Hey @najwalb - Taking over for Artsiom here. If you're using a specific integration, for example PyTorch/PTL, you can automatically save the best k models within the [ModelCheckpoint() callback](https://lightning.ai/docs/pytorch/stable/api/pytorch_lightning.callbacks.ModelCheckpoint.html#pytorch_lightning.callbacks.ModelCheckpoint) via the save_top_k argument and modify that callback to include saving this model as an artifact.

Without an integration, you'd like to save a model every time the best one improves, I would recommend introducing some logic like the following into your workflow after each epoch:
 
Let's say the 'best' model is determined by the highest accuracy for a given epoch, we would have the following steps:
1. Save your accuracy for each epoch and check to see if the current epoch has the best accuracy (or whichever condition suits your workflow)
2. If that model has the best accuracy, delete the existing version of the 'best model' saved to wandb 
3. If the current epoch has the best accuracy, save that model as an artifact 
4. Continue iterating until training is finished 
 
In order to delete artifacts programmatically within your workflow, you will need to utilize our API, and you will need the run_id of the current training run, project name, and entity. Here's sample code for how to delete all artifacts of type 'model' - but you can adjust this to your workflow:
 
```
api = wandb.Api()
run = api.run("<entity>/<project>/<run_id_of_current_run>")

for artifact in run.logged_artifacts():
    if artifact.type == "dataset":
        artifact.delete(delete_aliases=True)
```
Please let me know if this helps or if you have other questions!

**Comment by umakrishnaswamy** [comment]
Hi @najwalb, since we have not heard back from you we are going to close this request. If you would like to re-open the conversation, please let us know!

**Comment by sydholl** [comment]
We are closing this issue due to inactivity. If you feel that it merits additional review, please provide a reproduction and add a comment explaining why. Thank you.
